h2/src/test/org/h2/test/db/TestIndex.java

```diff
@@ -702,4 +707,29 @@ private void log(String sql) throws SQLException {
         trace("---done---");
     }
 
+    public static ResultSet testFunctionIndexFunction() {
+        SimpleResultSet rs = new SimpleResultSet();
+        rs.addColumn("ID", Types.INTEGER, ValueInt.PRECISION, 0);
+        rs.addColumn("VALUE", Types.INTEGER, ValueInt.PRECISION, 0);
+        rs.addRow(1, 10);
+        rs.addRow(2, 20);
+        rs.addRow(3, 30);
+        return rs;
+    }
+
+    private void testFunctionIndex() throws SQLException {
+        stat.execute("CREATE ALIAS TEST_INDEX FOR \"" + TestIndex.class.getName() + ".testFunctionIndexFunction\"");
+        try (ResultSet rs = stat.executeQuery("SELECT * FROM TEST_INDEX() WHERE ID = 1 OR ID = 3")) {
+            assertTrue(rs.next());
```

Comment:
this test is not verifying that we call testFunctionIndexFunction only once? which was the original bug.

Reply:
Fixed, but testing code now walks over thread stack to check who called the test function. Because this function is called at least twice even before execution of select statement.

Reply:
would using a PreparedStatement help with that?

Reply:
yeah, what is happening is that the engine needs to determine the metadata i.e. names and types of columns in resultset. Ideally we need a way for these kinds of functions to advertise their metadata separately from actually being executed

Reply:
Unfortunately my test is not stable, it fails in lazy mode.

Reply:
`PreparedStatement` does not help because `CommandContainer.recompileIfRequired()` is invoked anyway.

I slightly adjusted thread stack check so now test should update counter correctly in lazy mode too.